feat(kernels): implement polygon triangulation routines

imrn99 opened the pull request:
Implement two polygon triangulation methods:
- ear clipping, which works on all polygons with no hole
- fanning, which only works on a subset of polygons; there is a generic implementation of this method and one dedicated to convex polygons

## Scope

- [x] Code

## Type of change

- [x] New feature(s)

cedricchevalier19 commented on honeycomb-kernels/src/triangulation/ear_clipping.rs:
```diff
@@ -0,0 +1,134 @@
+        return;
+    }
+    if (n - 3) * 2 != new_darts.len() {
+        println!("W: not enough pre-allocated darts to triangulate face {face_id} -- skipping triangulation");
```
Is it always "not enough"?

cedricchevalier19 commented on honeycomb-kernels/src/triangulation/fan.rs:
```diff
@@ -0,0 +1,176 @@
+        return;
+    }
+    if (n - 3) * 2 != new_darts.len() {
+        println!("W: not enough pre-allocated darts to triangulate face {face_id} -- skipping triangulation");
+        return;
```
Same here.

cedricchevalier19 commented on honeycomb-kernels/src/triangulation/fan.rs:
```diff
@@ -0,0 +1,176 @@
+    new_darts: &[DartIdentifier],
+) {
+    // fetch darts using a custom orbit so that they're ordered
+    let darts: Vec<_> =
+        Orbit2::new(cmap, OrbitPolicy::Custom(&[1]), face_id as DartIdentifier).collect();
+    let n = darts.len();
+
+    // early rets
+    if n <= 3 {
+        println!("I: face {face_id} is an {n}-gon -- skipping triangulation");
+        return;
+    }
+    if (n - 3) * 2 != new_darts.len() {
+        println!("W: not enough pre-allocated darts to triangulate face {face_id} -- skipping triangulation");
+        return;
+    }
+
+    let vertices: Vec<_> = darts
+        .iter()
+        .map(|dart_id| {
+            cmap.vertex(cmap.vertex_id(*dart_id))
+                .expect("E: found a topological vertex with no associated coordinates")
+        })
+        .collect();
```
Duplicated code with `ear_clipping.rs`. Does a function make sense?

cedricchevalier19 commented on honeycomb-kernels/src/triangulation/fan.rs:
```diff
@@ -0,0 +1,176 @@
+) {
+    let n = Orbit2::new(cmap, OrbitPolicy::Custom(&[1]), face_id as DartIdentifier).count();
+
+    // early rets
+    if n <= 3 {
+        println!("I: face {face_id} is an {n}-gon -- skipping triangulation");
+        return;
+    }
+    if (n - 3) * 2 != new_darts.len() {
+        println!("W: not enough pre-allocated darts to triangulate face {face_id} -- skipping triangulation");
+        return;
+    }
```
Same here.

cedricchevalier19 reviewed (commented):
Nothing is blocking the PR, but I think you can improve the code before merging.

imrn99 commented:
@cedricchevalier19 I added three functions:
- `check_requirements`: this takes care of some of the duplicated code (number of vertex per face, number of pre-allocated darts)
- `fetch_face_vertices`: if all vertices of a face are defined, return a collection with those, else raise en error
- `crossp_from_verts`: this replace the macro; it should be inlined anyway

I also added the `TriangulateError` to describe these issues. The routines now return a `Result<(), TriangulateError>` that can be processed to shut down the program or attempt triangulation with another method.

cedricchevalier19 reviewed (approved):
LGTM
